sw/host/opentitanlib/src/io/i2c.rs

```diff
@@ -64,14 +66,69 @@ pub enum I2cError {
 }
 impl_serializable_error!(I2cError);
 
-/// Represents a I2C transfer.
+/// Represents a I2C transfer to be initiated by the debugger as I2C host.
 pub enum Transfer<'rd, 'wr> {
     Read(&'rd mut [u8]),
     Write(&'wr [u8]),
 }
 
+/// Status of I2C read operations (data from device to host).
+#[derive(Debug, Deserialize, Serialize)]
+pub enum ReadStatus {
+    /// Host has asked to read data, debugger device is currently stretching the clock waiting to
+    /// be told what data to transmit via I2C.  Parameter is 7-bit I2C address.
+    WaitingForData(u8),
+    /// Host is not asking to read data, and debugger also does not have any prepared data.
+    Idle,
+    /// Host is not asking to read data, debugger has prepared data for the case the host should
+    /// ask in the future.
+    DataPrepared,
+}
+
+/// Record of one transfer initiated by the I2C host, to which the debugger responded as I2C
+/// device.
+#[derive(Debug, Deserialize, Serialize)]
+pub enum DeviceTransfer {
+    /// The I2C host read a number of previously prepared bytes.
+    Read {
+        addr: u8,
+        /// True if `prepare_read_data()` was not called in time.
+        timeout: bool,
+        len: usize,
+    },
+    /// The I2C host wrote the given sequence of bytes.
+    Write { addr: u8, data: Vec<u8> },
+}
+
+/// A log of I2C operations performed by the I2C host since last time, as well as whether the I2C
+/// host is currently waiting to read data.
+#[derive(Debug, Deserialize, Serialize)]
+pub struct DeviceStatus {
+    /// Log of transfers completed since the last time.
+    pub transfers: Vec<DeviceTransfer>,
+    /// Indicates whether the I2C host is currently waiting to read data (clock stretched by
+    /// debugger).  If that is the case, the `tranfers` field above is guaranteed to contain all
+    /// write (and read) transfers performed before the currently blocked read transfer (unless
+    /// they have been already retrieved by a previous call to `get_device_status()`).
+    pub read_status: ReadStatus,
+}
+
 /// A trait which represents a I2C Bus.
 pub trait Bus {
+    /// Put I2C debugger into host mode (this is the default).
+    fn set_mode_host(&self) -> Result<()> {
+        Ok(())
+    }
+
+    /// Put I2C debugger into device mode.
+    fn set_mode_device(&self, _addr: u8) -> Result<()> {
+        Err(TransportError::UnsupportedOperation.into())
+    }
```

Comment:
Should we collapse this into just `fn set_mode(&self, Mode)`?

We can define the `Mode` enum to allow specifying the address:
```
pub enum Mode {
  Host,
  Device(u8),
}
```

Reply:
That makes sense. Done.

Reply:
This also makes it a little more convenient if in the future we want to support `DeviceWithAddressMask{addr: u8, mask: u8}`, or some other variation taking two separate addresses or something.